### ISSUE TITLE
How to test the bot with automatic tests?

### ISSUE THREAD

**Issue by invasionofsmallcubes** [yagop/node-telegram-bot-api#445]
> How to test the bot with automatic tests?

Hello, my question could be a little bit dull, but it could because I'm approaching now to the problem for the first time.
I would like to test that, e.g. `bot.onText('/\/ping/')`, it's triggering the correct use case. My idea would be to trigger the `processUpdate` method on `TelegramBot`.
The problem is that I always need to construct the `TelegramBot` with a token that does an actual call. Do you think there is a way in which I can avoid to mock an http call?

**Comment by invasionofsmallcubes** [comment]
So I finally solved the issue. Sort of. I used `jest` + `nock` and I was able to reproduce a polling condition. Although without `nock` I think it's impossible to test it.

**Comment by mezgoodle** [comment]
@invasionofsmallcubes 
Hello. Can you give me link to your repo(code) for showing testing? I am having similar problem now.

**Comment by invasionofsmallcubes** [comment]
> @invasionofsmallcubes
> Hello. Can you give me link to your repo(code) for showing testing? I am having similar problem now.

Yes, of course. But it’s a really old source code now. I haven’t been working on it for years. 

It’s the crypto-monitor project.

**Comment by mezgoodle** [comment]
> 
> 
> > @invasionofsmallcubes
> > Hello. Can you give me link to your repo(code) for showing testing? I am having similar problem now.
> 
> Yes, of course. But it’s a really old source code now. I haven’t been working on it for years.
> 
> It’s the crypto-monitor project.

Thank you

**Comment by aycanadal** [comment]
Just in case someone needs, I test like this without nock:

`const botService = app.get(BotService);
      botService.createBot();
      const onText = jest.spyOn(botService.bot, 'onText');
      await botService.startPolling();

      const innerOnText = onText.mock.calls[0][1]
      let receiverId = 453452319, userInput = `/adminhello ${receiverId} test`;

      let sendMessageArgs = { receiverId: null, forwardedMessage: null }

      const bot = {
        innerOnText: innerOnText, bot: {
          sendMessage: (receiverIdInner, forwardedMessageInner) => {
            sendMessageArgs.receiverId = receiverIdInner;
            sendMessageArgs.forwardedMessage = forwardedMessageInner
          }
        }
      }

      await bot.innerOnText(
        {
          text: userInput,
          message_id: 1,
          date: 123,
          chat: { id: 1, type: "private" },
          from: { id: 453452319, is_bot: false, first_name: "test" }
        },
        /\/adninhello /.exec(userInput))

        expect(sendMessageArgs.receiverId).toBe(receiverId)`

Basically I spy on the onText, start the bot, capture the second argument onText is called with. That is the callback. And then call the callback manually, and expect it to call sendMessage with correct arguments.
